### ISSUE TITLE
Full analytical derivatives in VTPR

### ISSUE THREAD

**Issue by JonWel** [CoolProp/CoolProp#1310]
This will not be easy and it will take times.

Right now, as time is limited, I'm thinking about doing one level at a time.

**Comment by JonWel** [comment]
First level from numerical derivatives of gE_R to a numerical derivative of ln_gamma_R:
![capture du 2016-11-02 17-01-09](https://cloud.githubusercontent.com/assets/11057909/19936546/121ec634-a11e-11e6-8d29-2131d86fee6d.png)

**Comment by ibell** [comment]
This is awesome.  I know it is going to be painful.  I'd be interested to
help; I haven't done crazy derivatives in a while...

**Comment by JonWel** [comment]
The most painful part will be the one for lnGamma, it has a fraction (and a ln which becomes a fraction after a derivative). Fractions gives big amount of terms: https://www.wolframalpha.com/input/?i=d%5E4(f(x)%2Fg(x))%2Fdx%5E4.
The remaining is still "nice", as you can see.

**Comment by JonWel** [comment]
After #1312, I think it will be required to change the structure of the code in order to introduce the other derivatives (we don't want to compute all derivatives when needing only the first one).

Basically, the `set_mole_fractions` remains, but the `set_temperature`, will change as the derivatives will need to be applied through it and we need to compute only the derivative of interest for a given call.
And if possible to cache as much as we can between calls (for instance, psi will be required several times for all the derivatives, and the derivatives of `ln_gamma_R` needs derivatives of different orders, so that may be a good idea to store a matrix of psi_mn in order to compute them only once per given temperature. This makes one psi_mn matrix for each fluids, so this will take a bit more memory).

I noticed `ln_Gamma_C` is computed at each `set_mole_fractions`, but it's only necessary in computing the `activity_coefficient`.
A first easy step may be to put it in a stand-alone function or in the `activity_coefficient` code, so this will slightly speed the other calls.
Be aware, if you are using `activity_coefficient` in your codes, that I'm changing the input arguments, in order to have it compatible with the new way of calling `ln_gamma_R`.

**Comment by JonWel** [comment]
I noticed `ln_Gamma_C` is computed at each `set_mole_fractions`, but it's only necessary in computing the `activity_coefficient`.

Shouldn't we put it in a stand-alone function or in the `activity_coefficient` code, so this will slightly speed the other calls?

The `activity_coefficient` function depends on the molar fraction and the temperature but currently relies on them to be set before.
How fast should that function be? (it's currently not used in other functions in CoolProp).
It currently gives back the coefficients one by one, which means that if the `ln_Gamma_R` and the `ln_Gamma_C` are computed completely at each call, there is way too much work done.
If we change to output all of them, this is already better.

I would see something like this:
```C++
void UNIFAQ::UNIFAQMixture::activity_coefficients(const double tau, const std::vector<double> &z, std::vector<double> &coefficients){
    std::size_t N = z.size();
    std::vector<double> r, q, l, phi, theta, ln_Gamma_C;
    double summerzr = 0, summerzq = 0, summerzl = 0;
    for (std::size_t i = 0; i < N; ++i) {
        double summerr = 0, summerq = 0;
        const UNIFAQLibrary::Component &c = components[i];
        for (std::size_t j = 0; j < c.groups.size(); ++j) {
            const UNIFAQLibrary::ComponentGroup &cg = c.groups[j];
            summerr += cg.count*cg.group.R_k;
            summerq += cg.count*cg.group.Q_k;
        }
        r[i] = summerr;
        q[i] = summerq;
        summerzr += z[i] * r[i];
        summerzq += z[i] * q[i];
    }
    for (std::size_t i = 0; i < N; ++i) {
        phi[i] = z[i] * r[i] / summerzr;
        theta[i] = z[i] * q[i] / summerzq;
        l[i] = 10.0 / 2.0*(r[i] - q[i]) - (r[i] - 1);
        summerzl += z[i] * l[i];
    }
    for (std::size_t i = 0; i < N; ++i) {
        ln_Gamma_C[i] = log(phi[i] / z[i]) + 10.0 / 2.0*q[i] * log(theta[i] / phi[i]) + l[i] - phi[i] / z[i] * summerzl;
         coefficients[i] = exp(ln_gamma_R(tau, i, 0) + ln_Gamma_C[i]);
    }
}
```

What do you think about it? (you know better than me what you would like to do with this function).

My main goal by doing this is to remove this part of the code from the `set_mole_fractions` to speed the saturation loop computation (PQ and QT loop change the molar fraction a lot, so the lighter  `set_mole_fractions` is, the faster they will be).

**Comment by ibell** [comment]
Yes, these are all good ideas.  My experience has not made it very clear whether using the Eigen arrays is necessarily faster (maybe, but maybe not), but it is certainly much easier to read that code!  I think (though we need to verify) that the additional memory storage for psi_mn matrix should not be very expensive.  

Let's do these pieces little by little and add tests as we go.

It would be very nice to come up with a better way of caching calculations so that we don't have to do them all the time.  How to decide *what* to cache is a harder question, but one worth thinking about.   But for instance your change to the activity coefficient is interesting.  Does that function actually get called in the code anywhere?  For VTPR, doesn't it just call the residual part of the activity coefficient?

**Comment by JonWel** [comment]
> Does that function actually get called in the code anywhere? For VTPR, doesn't it just call the residual part of the activity coefficient?

The only place it's called is in the test function. The VTPR does in deed only need the residual part.
This is precisely why I want to remove anything related to the comminatory part from the residual part computation. I'm currently computing it millions of times without doing anything with it.

> My experience has not made it very clear whether using the Eigen arrays is necessarily faster (maybe, but maybe not)

It may be a bit slower when things like `-msse2` are not enabled. There may be some "wrapping" overload and I guess each command may have it's own loop instead of assemble what can be.
When `-msse2` is active, this may not be that obvious either as it may already be optimised by the compiler.
So yes, this is not sure if this will improve speed.
I think the part that will really gives a boost is saving the right variables at the right moment. Not obvious to do, but I think there is definitely some room for improvement.

**Comment by ibell** [comment]
Yes, it is a good idea to not do calculations you don't need to - that is the core problem we have in VTPR right now - lots of calculations that are duplicative.  I think coming up with a good scheme for caching is important, and will potentially speed up HEOS quite a bit too if we arrive at a nice solution for VTPR.  For HEOS, my basic idea is to cache alphar, all first partial derivatives of alphar with respect to tau, delta for all mole fractions, and then other things that need these derivatives will hit the cache if needed, or will fill in the cache otherwise.  It would be a fair bit of memory, but not unmanageable I hope.  I might have a go at this rather soon, I've been mulling this over in my head for a long time.

**Comment by JonWel** [comment]
@ibell In case you didn't saw, I tested #1310 and saw no issue.

**Comment by JonWel** [comment]
I added the changes to `activity_coefficients` as proposed.

**Comment by JonWel** [comment]
Thinking about it again, eigen would be worth it for big matrices, where the wrapping overhead becomes quite negligible and optimisations possible, but in our case, we have really small arrays, and the AVX is already used if the flag is given to the compiler. Most of the other optimisations would have no impact on such small arrays while the eigen's wrapping will probably provides some overhead.

This will probably be faster to work with the default C++ arrays and for loops as currently. I think speed is a quite important aspect here, even if a bit less nice to read.

**Comment by JonWel** [comment]
On a 2 component mixture, the changes to `activity_coefficients` gives me a 2% time reduction on QT_INPUTS.
Not that much, but I think it's mainly because most of the time is spent in calling psi. Once the call to psi will be faster, this change will have been more worth it.

**Comment by JonWel** [comment]
@ibell what's the advantage of using references to some "arrays", instead of directly using the array. For instance:
```c++
std::map<std::size_t, double> &Xg = m_Xg, &thetag = m_thetag;
Xg.clear(); thetag.clear();
// (...)
    Xg.insert(std::pair<std::size_t, double>(it->sgi, X));
// (...)
```
Compared to:
```c++
m_Xg.clear(); m_thetag.clear();
// (...)
    m_Xg.insert(std::pair<std::size_t, double>(it->sgi, X));
// (...)
```

Because both will modify the same entities `m_Xg` and `m_thetag`, so why creating these temporary references? (to me, it creates two more mov instructions).

**Comment by ibell** [comment]
The point was only clarity, you are completely correct that those two bits are equivalent.  It's nice to see ``Xg`` instead of ``m_Xg`` is all.

**Comment by JonWel** [comment]
Ok, so I may use `Xg` everywhere as well.

**Comment by JonWel** [comment]
Mhh, I wrongly assumed there were enough derivative checks in the current catch, but after checking, there is only one composition derivative check. Sorry, I did not check well enough :s .
I added some on my repo (based on the testing file you sent me some times ago) and I have these values that are not considered good enough:
```
VTPR test
  dalphar, idelta = 0; itau = 1
-------------------------------------------------------------------------------
E:\GitHub\CoolProp-f\src\Backends\Cubics\VTPRBackend.cpp(130)
...............................................................................

E:\GitHub\CoolProp-f\src\Backends\Cubics\VTPRBackend.cpp(179): FAILED:
  CHECK( std::abs((numeric / analytic - 1) * 100) < 1e-7 )
with expansion:
  0.0000161025 < 0.0000001

-------------------------------------------------------------------------------
VTPR test
  dalphar, idelta = 0; itau = 2
-------------------------------------------------------------------------------
E:\GitHub\CoolProp-f\src\Backends\Cubics\VTPRBackend.cpp(130)
...............................................................................

E:\GitHub\CoolProp-f\src\Backends\Cubics\VTPRBackend.cpp(179): FAILED:
  CHECK( std::abs((numeric / analytic - 1) * 100) < 1e-7 )
with expansion:
  0.0000106328 < 0.0000001

-------------------------------------------------------------------------------
VTPR test
  dalphar, idelta = 0; itau = 3
-------------------------------------------------------------------------------
E:\GitHub\CoolProp-f\src\Backends\Cubics\VTPRBackend.cpp(130)
...............................................................................

E:\GitHub\CoolProp-f\src\Backends\Cubics\VTPRBackend.cpp(179): FAILED:
  CHECK( std::abs((numeric / analytic - 1) * 100) < 1e-7 )
with expansion:
  0.0000128202 < 0.0000001

-------------------------------------------------------------------------------
VTPR test
  dalphar, idelta = 0; itau = 4
-------------------------------------------------------------------------------
E:\GitHub\CoolProp-f\src\Backends\Cubics\VTPRBackend.cpp(130)
...............................................................................

E:\GitHub\CoolProp-f\src\Backends\Cubics\VTPRBackend.cpp(179): FAILED:
  CHECK( std::abs((numeric / analytic - 1) * 100) < 1e-7 )
with expansion:
  0.0000560566 < 0.0000001
```
So the derivatives I changed do not seem to be as close to the numerical as expected.

I however do not see what would be wrong in the way I proceeded.

**Comment by ibell** [comment]
Well, this problem might be caused by truncation error based on the size of
dtau in our derivatives.  Could you try playing with that and see if the
tests pass?

**Comment by ibell** [comment]
Closing as part of an issue-tracker cleanup — this issue has had no activity in over two years.

*To be transparent: this comment and the close action are being performed by Claude (Anthropic's AI coding assistant), running on behalf of @ibell. Ian reviewed and approved the bulk-close criteria; I am the one executing the action.*

If you would like this issue reconsidered, please feel free to **re-open it** — a fresh comment with current behaviour against the latest CoolProp release is the most useful way to revive the discussion.

— Claude (acting for @ibell)
